docs/adrs/006-web-apis.md

```diff
@@ -111,80 +108,61 @@ import type { PenumbraService } from '@penumbra-zone/protobuf';
-  readonly disconnect: () => Promise<void>;
-  /** Reexports the `isConnected` function from injected provider  */
-  readonly isConnected: () => boolean | undefined;
-  /** Reexports the `state` function from injected provider */
-  readonly getState: () => PenumbraState;
-  /** Reexports the `onConnectionChange` listener from injected provider*/
-  readonly onConnectionChange: (
-    cb: (connection: { origin: string; connected: boolean; state: PenumbraState }) => void,
-  ) => void;
-
-  /**
-   * Needed for custom service connections if `getService` is not enough.
-   * For example, might be useful for React wrapper of the `client` package
-   */
-  readonly getMessagePort: () => MessagePort;
-}
-```
 
-Moreover, the `client` package might export but not limited to the following useful functions and types:
+  // static methods
+  public static providerManifests: () => Record<string, Promise<PenumbraManifest>>;
 
-```ts
-export type PenumbraManifest = Partial<chrome.runtime.ManifestV3> &
-  Required<Pick<chrome.runtime.ManifestV3, 'name' | 'version' | 'description' | 'icons'>>;
+  // provider-specific static methods
+  public static providerManifest: (providerOrigin?: string) =>  Promise<PenumbraManifest>;
+  public static providerIsConnected: (providerOrigin?: string) => boolean;
+  public static providerState: (providerOrigin?: string) => PenumbraState;
 
-export type getInjectedProvider = (penumbraOrigin: string) => Promise<PenumbraProvider>;
+  /** Initiates initiates connection, returns a client instance. */
+  public static providerConnect: (providerOrigin?: string) => Promise<PenumbraClient>;
 
-export type getAllInjectedProviders = () => string[];
+  // instance features
 
-export type getPenumbraManifest = (
-  penumbraOrigin: string,
-  signal?: AbortSignal,
-) => Promise<PenumbraManifest>;
+  /** Creates a client instance for a provider origin but does not take any action. */
+  constructor(providerOrigin?: string) => PenumbraClient
 
-export type getAllPenumbraManifests = () => Record<
-  keyof (typeof window)[typeof PenumbraSymbol],
-  Promise<PenumbraManifest>
->;
-```
+  // constructor input
+  public readonly providerOrigin: string;
 
-## Requests
+  // populated when client is in appropriate state.
+  public readonly transport?: Transport;
+  public readonly port?: MessagePort;
 
-The client library should separately export the creation of service clients. It is not going to be integrated into the `client` instance to save the initial bundle size. Instead, it should be exported from `@penumbra-zone/client/service`:
+  /** Direct re-exports the from appropriate provider */
+  public readonly disconnect: () => Promise<void>;
+  public readonly isConnected: () => boolean | undefined;
+  public readonly state: () => PenumbraState;
+  public readonly addEventListener: EventTarget['addEventListener'];
+  public readonly removeEventListener: EventTarget['removeEvenListener'];
```

Comment:
thought (possible complication): i'm still hesitant of this typing. It requires importing a helper function:

```ts
client.addEventListener('penumbrastate', (event) => {
  if (isPenumbraStateEvent(evt)) {
    // Do something
  }
});
```

This import makes the use way more complicated. Especially that most of the methods of the library come from the `PenumbraClient`, it makes it ambiguous.

suggestion: simplify the typings

Reply:
question: what happened to the `onConnectionChange`? It worked like a charm here: https://github.com/penumbra-zone/nextjs-penumbra-client-example/pull/5/files#diff-6e5a42ea42819066f6aa0ff0e0d91ef85f5b435355e6af424cc2e25e9d03a8f0R38

Reply:
in the minifront implementation, there were issues with race conditions in the callback. the event handler would actually fire, and then activate the callbacks in the map, before the new state was available in the client object.

i can re-add this

Reply:
a callback function is restored now